Sources/Core/Client/Client+RxEvents.swift

```diff
@@ -0,0 +1,125 @@
+    /// - Parameters:
+    ///   - eventTypes: event types.
+    ///   - channel: a channel.
+    /// - Returns: an observable channel events.
+    func onEvent(eventTypes: [EventType] = [], channel: Channel) -> Observable<StreamChatClient.Event> {
+        onEventConnected(eventTypes: eventTypes, channel: channel)
+    }
+}
+
+// MARK: Private Client Rx Events
+
+extension Client {
+    private static var rxOnTokenChange: UInt8 = 0
+    private static var rxOnConnect: UInt8 = 0
+    private static var rxOnEvent: UInt8 = 0
+    private static var rxOnUserUpdate: UInt8 = 0
+    
+    fileprivate var rxOnTokenChange: Observable<Token?> {
+        associated(to: self, key: &Client.rxOnTokenChange) { [unowned self] in
+            Observable<Token?>.create({ [unowned self] observer in
+                self.onTokenChange = { observer.onNext($0) }
+                return Disposables.create()
+            })
+                .startWith(token)
+                .share(replay: 1)
+        }
+    }
+    
+    fileprivate var rxOnConnect: Observable<Connection> {
+        associated(to: self, key: &Client.rxOnConnect) { [unowned self] in
+            Observable<Connection>.create({ [unowned self] observer in
+                self.onConnect = { observer.onNext($0) }
+                return Disposables.create()
+            })
+                .startWith(connection)
+                .share()
+        }
+    }
+    
+    fileprivate var rxOnEvent: Observable<StreamChatClient.Event> {
+        associated(to: self, key: &Client.rxOnEvent) { [unowned self] in
+            Observable<StreamChatClient.Event>.create({ observer in
+                self.onEvent = { observer.onNext($0) }
+                return Disposables.create()
+            }).share()
+        }
+    }
+    
+    fileprivate var rxOnUserUpdate: Observable<User> {
+        associated(to: self, key: &Client.rxOnUserUpdate) { [unowned self] in
+            Observable<StreamChatClient.User>.create({ observer in
+                self.onUserUpdate = { observer.onNext($0) }
+                return Disposables.create()
+            })
+                .startWith(user)
+                .share(replay: 1)
+        }
+    }
+}
+
+extension Reactive where Base == Client {
+    
+    var onTokenChange: Observable<Token?> { base.rxOnTokenChange }
+    var onConnect: Observable<Connection> { base.rxOnConnect }
+    var onEvent: Observable<StreamChatClient.Event> { base.rxOnEvent }
+    var onUserUpdate: Observable<User> { base.rxOnUserUpdate }
+    
+    func onEventConnected(eventTypes: [EventType], channel: Channel? = nil) -> Observable<StreamChatClient.Event> {
+        connection
+            .filter({ $0.isConnected })
+            .flatMapLatest { _ in self.onEvent }
+            .filter({ [weak channel] in
+                if !eventTypes.isEmpty, !eventTypes.contains($0.type) {
+                    return false
+                }
+                
+                if let cid = channel?.cid {
```

Comment:
`return channel?.cid == $0.cid` should be enough, no need for `if let`

Reply:
Almost. I've updated, but if the channel is nil, no need to check `$0.cid`.

Reply:
Compiler will optimize it anyway, if client is `nil` `$0.cid` won't be checked

Reply:
This is the latest and right condition. It's not about compiler optimization. The logic will be different. The channel value is opitonal.
```swift
eventTypes.isEmpty || (eventTypes.contains($0.type) && (channel == nil || $0.cid == channel?.cid))
```

Reply:
Now I made a mistake with optimization. the original implementation was correct. The `eventTypes` is optional too.
Cases:
1. Both values empty
2. `eventTypes` is not empty, `channel` is nil
3. `eventTypes` is not empty, `channel` is not nil
4. `eventTypes` is empty, `channel` is not nil

Reply:
I moved the implementation back.

Reply:
Ok, at this point we don't need `if let`, we can just `return channel?.cid == $0.cid` and it will not read `$0.cid` if `channel` is `nil`, since rhs is not optional and lhs is.

Reply:
`rhs` is optional too.